File: study-flink-1.14.2/src/main/java/com/yee/study/bigdata/flink114/java/table/TableApiSample.java
package com.yee.study.bigdata.flink114.java.table;

import org.apache.flink.api.common.RuntimeExecutionMode;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;

import java.time.Instant;

import static org.apache.flink.table.api.Expressions.$;
import static org.apache.flink.table.api.Expressions.concat;

/**
 * @author Roger.Yi
 */
public class TableApiSample {

    public static void main(String[] args) throws Exception {
        // Stream env
        StreamExecutionEnvironment sEnv = StreamExecutionEnvironment.getExecutionEnvironment();
        sEnv.setRuntimeMode(RuntimeExecutionMode.STREAMING);

        // Table env
        StreamTableEnvironment tabEnv = StreamTableEnvironment.create(sEnv);

        // Scan, Projection, Filter 示例
//        scanSamples(sEnv, tabEnv);

        // Column Operation 示例
//        columnOperationSamples(sEnv, tabEnv);

        // Aggregation 示例
        aggregationSamples(sEnv, tabEnv);
    }

    /**
     * Scan, Projection, Filter 示例
     *
     * @param sEnv
     * @param tabEnv
     * @throws Exception
     */
    public static void scanSamples(StreamExecutionEnvironment sEnv, StreamTableEnvironment tabEnv) throws Exception {
        // DataStream
        DataStream<Order> dataStream = sEnv.fromElements(
                new Order("Roger", "China", "Shanghai", Instant.ofEpochMilli(1000)),
                new Order("Andy", "China", "Shanghai", Instant.ofEpochMilli(1000)),
                new Order("John", "USA", "NY", Instant.ofEpochMilli(1000)),
                new Order("Eric", "USA", "LA", Instant.ofEpochMilli(1000)),
                new Order("Rooney", "UK", "MAN", Instant.ofEpochMilli(1000)));

        tabEnv.createTemporaryView(
                "Orders",
                dataStream);
        Table orders = tabEnv.from("Orders"); // schema (a, b, c, rowtime)

        /**
         * 1) Overview & Samples
         * +----+--------------------------------+----------------------+
         * | op |                              a |                  cnt |
         * +----+--------------------------------+----------------------+
         * | +I |                          Roger |                    1 |
         * | +I |                         Rooney |                    1 |
         * | +I |                           Andy |                    1 |
         * | +I |                           John |                    1 |
         * | +I |                           Eric |                    1 |
         * +----+--------------------------------+----------------------+
         */
        Table counts = orders
                .groupBy($("a"))
                .select($("a"), $("b").count().as("cnt"));
        counts.execute().print();

        /**
         * 2) Select
         * +----+--------------------------------+--------------------------------+
         * | op |                              a |                              d |
         * +----+--------------------------------+--------------------------------+
         * | +I |                          Roger |                       Shanghai |
         * | +I |                           Andy |                       Shanghai |
         * | +I |                           John |                             NY |
         * | +I |                           Eric |                             LA |
         * | +I |                         Rooney |                            MAN |
         * +----+--------------------------------+--------------------------------+
         */
        Table result = orders.select($("a"), $("c").as("d")); // orders.select($("*"));
        result.execute().print();

        /**
         * 3) As
         * +----+----------+----------+-------------+-------------------------------+
         * | op |        x |        y |           z |                             t |
         * +----+----------+----------+-------------+-------------------------------+
         * | +I |    Roger |    China |    Shanghai | 1970-01-01 08:00:01.000000000 |
         * | +I |     Andy |    China |    Shanghai | 1970-01-01 08:00:01.000000000 |
         * | +I |     John |      USA |          NY | 1970-01-01 08:00:01.000000000 |
         * | +I |     Eric |      USA |          LA | 1970-01-01 08:00:01.000000000 |
         * | +I |   Rooney |       UK |         MAN | 1970-01-01 08:00:01.000000000 |
         * +----+----------+----------+-------------+-------------------------------+
         */
        result = orders.as("x, y, z, t");
        result.execute().print();

        /**
         * 4) Where/Filter
         *
         * +----+--------+--------+-----------+-------------------------------+
         * | op |      a |      b |         c |                       rowtime |
         * +----+--------+--------+-----------+-------------------------------+
         * | +I |  Roger |  China |  Shanghai | 1970-01-01 08:00:01.000000000 |
         * | +I |   Andy |  China |  Shanghai | 1970-01-01 08:00:01.000000000 |
         * +----+--------+--------+-----------+-------------------------------+
         */
        result = orders.filter($("b").isEqual("China"));
        result.execute().print();
    }

    /**
     * Column Operator 示例
     *
     * @param sEnv
     * @param tabEnv
     * @throws Exception
     */
    public static void columnOperationSamples(StreamExecutionEnvironment sEnv, StreamTableEnvironment tabEnv) throws Exception {
        // DataStream
        DataStream<Order> dataStream = sEnv.fromElements(
                new Order("Roger", "China", "Shanghai", Instant.ofEpochMilli(1000)),
                new Order("Andy", "China", "Shanghai", Instant.ofEpochMilli(1000)),
                new Order("John", "USA", "NY", Instant.ofEpochMilli(1000)),
                new Order("Eric", "USA", "LA", Instant.ofEpochMilli(1000)),
                new Order("Rooney", "UK", "MAN", Instant.ofEpochMilli(1000)));

        tabEnv.createTemporaryView(
                "Orders",
                dataStream);
        Table orders = tabEnv.from("Orders"); // schema (a, b, c, rowtime)

        /**
         * 1) AddColumns
         * +----+----------+---------+------------+-------------------------------+-------------------+
         * | op |        a |       b |          c |                       rowtime |              desc |
         * +----+----------+---------+------------+-------------------------------+-------------------+
         * | +I |    Roger |   China |   Shanghai | 1970-01-01 08:00:01.000000000 |     Shanghaisunny |
         * | +I |     Andy |   China |   Shanghai | 1970-01-01 08:00:01.000000000 |     Shanghaisunny |
         * | +I |     John |     USA |         NY | 1970-01-01 08:00:01.000000000 |           NYsunny |
         * | +I |     Eric |     USA |         LA | 1970-01-01 08:00:01.000000000 |           LAsunny |
         * | +I |   Rooney |      UK |        MAN | 1970-01-01 08:00:01.000000000 |          MANsunny |
         * +----+----------+---------+------------+-------------------------------+-------------------+
         */
        Table result = orders.addColumns(concat($("c"), "sunny").as("desc"));
        result.execute().print();

        /**
         * 2) AddOrReplaceColumns
         */
        result = orders.addOrReplaceColumns(concat($("c"), "sunny").as("desc"));
        result.execute().print();

        /**
         * 3) DropColumns
         * +----+--------------------------------+-------------------------------+
         * | op |                              a |                       rowtime |
         * +----+--------------------------------+-------------------------------+
         * | +I |                          Roger | 1970-01-01 08:00:01.000000000 |
         * | +I |                           Andy | 1970-01-01 08:00:01.000000000 |
         * | +I |                           John | 1970-01-01 08:00:01.000000000 |
         * | +I |                           Eric | 1970-01-01 08:00:01.000000000 |
         * | +I |                         Rooney | 1970-01-01 08:00:01.000000000 |
         * +----+--------------------------------+-------------------------------+
         */
        result = orders.dropColumns($("b"), $("c"));
        result.execute().print();

        /**
         * 4) RenameColumns
         * +----+--------+-------+----------+-------------------------------+
         * | op |      a |    b2 |       c2 |                       rowtime |
         * +----+--------+-------+----------+-------------------------------+
         * | +I |  Roger | China | Shanghai | 1970-01-01 08:00:01.000000000 |
         * | +I |   Andy | China | Shanghai | 1970-01-01 08:00:01.000000000 |
         * | +I |   John |   USA |       NY | 1970-01-01 08:00:01.000000000 |
         * | +I |   Eric |   USA |       LA | 1970-01-01 08:00:01.000000000 |
         * | +I | Rooney |    UK |      MAN | 1970-01-01 08:00:01.000000000 |
         * +----+--------+-------+----------+-------------------------------+
         */
        result = orders.renameColumns($("b").as("b2"), $("c").as("c2"));
        result.execute().print();
    }

    /**
     * Aggregation 示例
     *
     * @param sEnv
     * @param tabEnv
     * @throws Exception
     */
    public static void aggregationSamples(StreamExecutionEnvironment sEnv, StreamTableEnvironment tabEnv) throws Exception {
        // DataStream
        DataStream<Order> dataStream = sEnv.fromElements(
                new Order("Roger", "China", "Shanghai", Instant.ofEpochMilli(1000)),
                new Order("Andy", "China", "Shanghai", Instant.ofEpochMilli(1000)),
                new Order("John", "USA", "NY", Instant.ofEpochMilli(1000)),
                new Order("Eric", "USA", "LA", Instant.ofEpochMilli(1000)),
                new Order("Rooney", "UK", "MAN", Instant.ofEpochMilli(1000)));

        tabEnv.createTemporaryView(
                "Orders",
                dataStream);
        Table orders = tabEnv.from("Orders"); // schema (a, b, c, rowtime)

        /**
         * 1) GroupBy
         * +----+--------------------------------+----------------------+
         * | op |                              a |                    d |
         * +----+--------------------------------+----------------------+
         * | +I |                           Eric |                    1 |
         * | +I |                          Roger |                    1 |
         * | +I |                         Rooney |                    1 |
         * | +I |                           Andy |                    1 |
         * | +I |                           John |                    1 |
         * +----+--------------------------------+----------------------+
         */
        Table result = orders.groupBy($("a")).select($("a"), $("b").count().as("d"));
        result.execute().print();
    }

    public static class Order {
        public String a;
        public String b;
        public String c;
        public Instant rowtime;

        public Order(String a, String b, String c, Instant rowtime) {
            this.a = a;
            this.b = b;
            this.c = c;
            this.rowtime = rowtime;
        }

        public Order() {
        }
    }
}

File: study-flink-1.16/src/main/scala/com/yee/study/bigdata/flink116/scala/TableApiSample.scala
package com.yee.study.bigdata.flink116.scala

import org.apache.flink.api.scala.createTypeInformation
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.table.api._
import org.apache.flink.table.api.bridge.scala.StreamTableEnvironment

import java.time.Instant

/**
 *
 * @author Roger.Yi
 */
object TableApiSample {

  def main(args: Array[String]): Unit = {
    //    sample_1() // 基本 Table Api 使用框架
    sample_2() // fromDataStream 示例
  }

  /**
   * 基本 Table Api 使用框架
   *
   * @param tableEnv
   */
  def sample_1(): Unit = {
    // Create Env
    val settings = EnvironmentSettings
      .newInstance()
      .build()
    val tableEnv = TableEnvironment.create(settings)

    // Create a source table
    tableEnv.createTemporaryTable("SourceTable",
      TableDescriptor.forConnector("datagen")
        .schema(
          Schema
            .newBuilder()
            .column("f0", DataTypes.STRING())
            .build())
        .option("number-of-rows", "3")
        .build())

    // Create a sink table (using SQL DDL)
    tableEnv.executeSql("CREATE TEMPORARY TABLE SinkTable WITH ('connector' = 'print') LIKE SourceTable (EXCLUDING OPTIONS) ")

    // Create a Table object from a Table API query
    val table1 = tableEnv.from("SourceTable")

    // Create a Table object from a SQL query
    val table2 = tableEnv.sqlQuery("SELECT * FROM SourceTable")

    // Emit a Table API result Table to a TableSink, same for SQL result
    val tableResult = table1.insertInto("SinkTable").execute().print()
  }

  /**
   * fromDataStream 示例
   *
   * @param sEnv
   * @param tEnv
   * @return
   */
  def sample_2(): Unit = {
    val sEnv = StreamExecutionEnvironment.getExecutionEnvironment
    val tEnv = StreamTableEnvironment.create(sEnv)

    val ds = sEnv.fromElements(
      User("Roger", 10, Instant.ofEpochMilli(1000)),
      User("Andy", 3, Instant.ofEpochMilli(1001)),
      User("John", 6, Instant.ofEpochMilli(1002)))

    /**
     * 1) fromDataStream(DataStream)
     * 根据自定义类型获取字段
     * (
     * `name` STRING,
     * `score` INT NOT NULL,
     * `event_time` TIMESTAMP_LTZ(9)
     * )
     */
    val table1 = tEnv.fromDataStream(ds)
    table1.printSchema()

    /**
     * 2) fromDataStream(DataStream, Schema)
     * 自动获取字段类型、并增加一个计算字段（基于proctime）
     * (
     * `name` STRING,
     * `score` INT NOT NULL,
     * `event_time` TIMESTAMP_LTZ(9),
     * `proc_time` TIMESTAMP_LTZ(3) NOT NULL *PROCTIME* AS PROCTIME()
     * )
     */
    val table2 = tEnv.fromDataStream(
      ds,
      Schema.newBuilder
        .columnByExpression("proc_time", "PROCTIME()")
        .build
    )
    table2.printSchema()

    /**
     * 3) fromDataStream(DataStream, Schema)
     * 自动获取字段类型、并增加一个计算字段（基于proctime）和一个自定义的watermark策略
     * (
     * `name` STRING,
     * `score` INT NOT NULL,
     * `event_time` TIMESTAMP_LTZ(9),
     * `rowtime` TIMESTAMP_LTZ(3) *ROWTIME* AS CAST(event_time AS TIMESTAMP_LTZ(3)),
     * WATERMARK FOR `rowtime`: TIMESTAMP_LTZ(3) AS rowtime - INTERVAL '10' SECOND
     * )
     */
    val table3 = tEnv.fromDataStream(
      ds,
      Schema.newBuilder
        .columnByExpression("rowtime", "CAST(event_time AS TIMESTAMP_LTZ(3))")
        .watermark("rowtime", "rowtime - INTERVAL '10' SECOND")
        .build
    )
    table3.printSchema()

    /**
     * 4) fromDataStream(DataStream, Schema)
     * 手动定义字段类型
     * (
     * `name` STRING,
     * `score` INT NOT NULL,
     * `event_time` TIMESTAMP_LTZ(9),
     * `rowtime` TIMESTAMP_LTZ(3) *ROWTIME* METADATA,
     * WATERMARK FOR `rowtime`: TIMESTAMP_LTZ(3) AS SOURCE_WATERMARK()
     * )
     */
    val table4 = tEnv.fromDataStream(
      ds,
      Schema.newBuilder
        .columnByMetadata("rowtime", "TIMESTAMP_LTZ(3)")
        .watermark("rowtime", "SOURCE_WATERMARK()")
        .build
    )
    table4.printSchema()
  }

  case class User(name: String, score: java.lang.Integer, event_time: java.time.Instant)

}
